lib/scheduler/manual.go

```diff
@@ -119,51 +157,378 @@ func (mec ManualExecutionConfig) GetExecutionRequirements(es *lib.ExecutionSegme
+	}
+}
+
+// Run constantly loops through as many iterations as possible on a variable
+// dynamically controlled number of VUs either for the specified duration, or
+// until the test is manually stopped.
+//
+//TODO: split this up? somehow... :/
+func (mex *ManualExecution) Run(parentCtx context.Context, out chan<- stats.SampleContainer) (err error) {
+	mex.configLock.RLock()
+	// Safely get the current config - it's important that the close of the
+	// hasStarted channel is inside of the lock, so that there are no data races
+	// between it and the UpdateConfig() method.
+	currentControlConfig := mex.currentControlConfig
+	close(mex.hasStarted)
+	mex.configLock.RUnlock()
+
+	segment := mex.executorState.Options.ExecutionSegment
+	duration := time.Duration(currentControlConfig.Duration.Duration)
+
+	ctx, cancel := context.WithCancel(parentCtx)
+	defer cancel()
+	if duration > 0 { // Only keep track of duration if it's not infinite
+		go mex.stopWhenDurationIsReached(ctx, duration, cancel)
+	}
+
+	mex.logger.WithFields(
+		logrus.Fields{"type": manualExecutionType, "duration": duration},
+	).Debug("Starting scheduler run...")
+
+	// Retrieve and initialize the (scaled) number of MaxVUs from the global VU
+	// buffer that the user originally specified in the JS config.
+	startMaxVUs := segment.Scale(mex.startConfig.MaxVUs.Int64)
+	vuHandles := make([]*manualVUHandle, startMaxVUs)
+	activeVUsCount := new(int64)
+	runIteration := getIterationRunner(mex.executorState, mex.logger, out)
+	for i := int64(0); i < startMaxVUs; i++ { // get the initial planned VUs from the common buffer
+		vu, vuGetErr := mex.executorState.GetPlannedVU(mex.logger, false)
+		if vuGetErr != nil {
+			return vuGetErr
+		}
+		vuHandle := newManualVUHandle(
+			parentCtx, mex.executorState, activeVUsCount, vu, mex.logger.WithField("vuNum", i),
+		)
+		go vuHandle.runLoopsIfPossible(runIteration)
+		vuHandles[i] = vuHandle
+	}
+
+	// Keep track of the progress
+	maxVUs := new(int64)
+	*maxVUs = startMaxVUs
+	progresFn := func() (float64, string) {
+		spent := mex.executorState.GetCurrentTestRunDuration()
+		progress := 0.0
+		if duration > 0 {
+			progress = math.Min(1, float64(spent)/float64(duration))
+		}
+		//TODO: simulate spinner for the other case or cycle 0-100?
+		currentActiveVUs := atomic.LoadInt64(activeVUsCount)
+		currentMaxVUs := atomic.LoadInt64(maxVUs)
+		vusFmt := pb.GetFixedLengthIntFormat(currentMaxVUs)
+		return progress, fmt.Sprintf(
+			"currently "+vusFmt+" out of "+vusFmt+" active looping VUs, %s/%s", currentActiveVUs, currentMaxVUs,
+			pb.GetFixedLengthDuration(spent, duration), duration,
+		)
+	}
+	mex.progress.Modify(pb.WithProgress(progresFn))
+	go trackProgress(parentCtx, ctx, ctx, mex, progresFn)
+
+	currentlyPaused := false
+	waitVUs := func(from, to int64) {
+		for i := from; i < to; i++ {
+			vuHandles[i].wg.Wait()
+		}
+	}
+	handleConfigChange := func(oldControlConfig, newControlConfig ManualExecutionControlConfig) error {
```

Comment:
This should be resolved by https://github.com/loadimpact/k6/pull/1007/commits/2ebe3dfdd6c3ef97decbd9ac80c6747dedb3aac1. I couldn't just add it as a new method to the executor, since it just depends on a ton of things that can only be initialized by `Run()`... So, that's why I created a new `runState` struct that `Run()` initializes and I moved this lambda as one of its methods. While it's slightly more verbose, I think in this way it's much cleaner and we can't confuse which parts of the executor are initialized when. As a bonus, the `progresFn` and initial VU retrieval from the common buffer could also be methods in that new struct :tada: